deepchem/models/logger.py

```diff
@@ -0,0 +1,64 @@
+from typing import Optional, Union, Dict, List
+import tensorflow as tf
+import torch
+import numpy as np
+from deepchem.models import Model
+
+numeric = Union[tf.Tensor, torch.Tensor, int, float, complex, np.number]
+tensor = Union[tf.Tensor, torch.Tensor]
+
+
+class Logger(object):
+
+  def __init__(self):
+    raise NotImplementedError
+
+  def __iter__(self):
+    return self
+
+  def setup(self, config: Dict):
+    raise NotImplementedError
+
+  def finish(self):
+    raise NotImplementedError
+
+  def log_batch(self,
+                loss: Dict,
+                step: int,
+                inputs: tensor,
+                labels: tensor,
+                location: Optional[str] = None,
```

Comment:
This is basically the `group` parameter from before. It's a string to uniquely identify where the logging is happening so when viewed in Wandb dashboard, the metric charts are under the correct headings (train/, callback1/, callback2/, etc).  Otherwise, you could potentially have a train and a callback both logging the same metric (for example loss) and the loggers wouldn't be able to distinguish. I figured renaming it to `location` sounds better since "group" isn't really meaningful.

Originally I was going to have the "location" param just for the WandbLogger but since we agreed that the API should be common and have the same parameters, I added it to the generic Logger API and made it optional.